### ISSUE TITLE
[STORM-216] Secure Multi-tenant Storm

### ISSUE THREAD

**Issue by jira-importer** [apache/storm#4058]
<p>Yahoo has working code for a secure multi-tenant storm at <a href="https://github.com/yahoo/incubator-storm/tree/security" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/yahoo/incubator-storm/tree/security</a></p>

<p>We should break the code up into individual pieces, review it and pull it in.</p>

<p>This is to track all of the different parts of that work.</p>

---
<details><summary><i>Originally reported by <a title="Robert Joseph Evans" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2">revans2</a>, imported from: <a href="https://issues.apache.org/jira/browse/STORM-216" target="_blank">Secure Multi-tenant Storm</a></i></summary>
<i><ul>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2014-11-13T19:41:45+00:00
<li><b>imported</b>: 2025-01-24
</ul></i>
</details>

**Comment by jira-importer** [comment]
Subtasks:

- STORM-217
- STORM-218
- STORM-219
- STORM-220
- STORM-221
- STORM-222
- STORM-223
- STORM-224
- STORM-225
- STORM-226
- STORM-227
- STORM-228
- STORM-229
- STORM-231
- STORM-232
- STORM-233
- STORM-234
- STORM-235

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2">revans2</a>:</i>
<p>Ww have run into a number of issues where ZK Authentication will not work properly with going to newer versions of Curator (1.3.3) and Zookeeper (3.4.5)</p>

<p>Also The current Kerberos login code uses internal ZK classes that exist in 3.4.5.  We may want to just copy the pieces we need directly instead though.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>GitHub user revans2 opened a pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/121" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/121</a></p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-216" title="Secure Multi-tenant Storm" class="issue-link" data-issue-key="STORM-216"><del>STORM-216</del></a>: Added Authentication and Authorization.</p>

<p>    This is an upmerged version of <a href="https://github.com/yahoo/incubator-storm/tree/security" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/yahoo/incubator-storm/tree/security</a></p>

<p>    Most of the documentation for this should be in SECURITY.md.  The documentation may not be perfect and if you have any questions while trying to set this up please ask them on the pull request so that I can clean up the documentation and make it complete.</p>

<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/apache/incubator-storm" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm</a> security</p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/121.patch" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/121.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #121</p>

<hr />
<p>commit fe5f41aa8332700d3f98422cb7d986fc47289bcd<br/>
Author: Robert (Bobby) Evans <bobby@apache.org><br/>
Date:   2014-05-21T16:03:11Z</p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-216" title="Secure Multi-tenant Storm" class="issue-link" data-issue-key="STORM-216"><del>STORM-216</del></a>: Added Authentication and Authorization.</p>

<p>    This is an upmerged version of <a href="https://github.com/yahoo/incubator-storm/tree/security" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/yahoo/incubator-storm/tree/security</a></p>

<hr />

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/121#issuecomment-44408880" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/121#issuecomment-44408880</a></p>

<p>    Just upmerged to the latest master, and pulled in a test fix.  Be aware that on a Mac (which has a case insensitive file system) the update may not go that cleanly.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/121#issuecomment-45763838" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/121#issuecomment-45763838</a></p>

<p>    Just upmerged to the latest master.  This was not that trivial because of the UI changes that switched over to using templates and AJAX.  The only part I did not bring over was displaying the logged in user on the UI page.  I am going to continue to work on this, and should hopefully have that up soon.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/121#issuecomment-45912041" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/121#issuecomment-45912041</a></p>

<p>    This pull request has been up for 3 weeks without a single comment yet.  I know it is a huge patch, but any feedback on it would be appreciated.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user miguno commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/121#issuecomment-47404034" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/121#issuecomment-47404034</a></p>

<p>    Bobby, do you know what the relation of this (big) ticket is to <a href="https://issues.apache.org/jira/browse/STORM-146" title="Unit test regression when storm is compiled with 3.4.5 zookeeper" class="issue-link" data-issue-key="STORM-146"><del>STORM-146</del></a>(<a href="https://issues.apache.org/jira/browse/STORM-146" class="external-link" rel="nofollow">https://issues.apache.org/jira/browse/STORM-146</a>), which is marked as "contained by" this ticket?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/121#issuecomment-50641173" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/121#issuecomment-50641173</a></p>

<p>    @miguno wow it took me a month to respond to your comment, I am really sorry.  I resolved <a href="https://issues.apache.org/jira/browse/STORM-146" title="Unit test regression when storm is compiled with 3.4.5 zookeeper" class="issue-link" data-issue-key="STORM-146"><del>STORM-146</del></a> originally pointing to this branch as the fix, because the tests all pass on ZK 3.4.5 here.  The fix is related to IPv6 timing out on the mac and causing issues.  When we added in a config to the pom to ask it to favor IPv4 when running the tests everything was fine.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2">revans2</a>:</i>
<p>I think this is just about ready to be pulled into master.  </p>

<p><a href="https://issues.apache.org/jira/issues/?jql=project%20%3D%20STORM%20AND%20resolution%20%3D%20Unresolved%20AND%20labels%20in%20%28%22security%22%29" class="external-link" rel="nofollow">https://issues.apache.org/jira/issues/?jql=project%20%3D%20STORM%20AND%20resolution%20%3D%20Unresolved%20AND%20labels%20in%20%28%22security%22%29</a></p>

<p>Lists all of the outstanding issues tagged with security and all of them except <a href="https://issues.apache.org/jira/browse/STORM-427" title="(Security) AutoTGT with HBase can expose JVM kerberos issue" class="issue-link" data-issue-key="STORM-427"><del>STORM-427</del></a> and <a href="https://issues.apache.org/jira/browse/STORM-408" title="Cross-Site Scripting security vulnerability" class="issue-link" data-issue-key="STORM-408"><del>STORM-408</del></a> are features that are not critical.  <a href="https://issues.apache.org/jira/browse/STORM-408" title="Cross-Site Scripting security vulnerability" class="issue-link" data-issue-key="STORM-408"><del>STORM-408</del></a> exists in master as well as security so I don't see it as a blocker.  If I could get a review for <a href="https://issues.apache.org/jira/browse/STORM-427" title="(Security) AutoTGT with HBase can expose JVM kerberos issue" class="issue-link" data-issue-key="STORM-427"><del>STORM-427</del></a> we can then pull it in and I think we will be ready.</p>

<p>We have been running in production with most of this code for quite a while.  If some others could take a look at this, try it out I would love to get this merged in the next week, so I can stop up-merging.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user miguno commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/121#issuecomment-50666328" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/121#issuecomment-50666328</a></p>

<p>    Thanks for the follow-up, Bobby.</p>

<p>    FYI:  I learned about the IPv6 timing issue on Mac OS X in the meantime, and I also updated the JVM configuration settings (i.e. disabling IPv6) in Storm-based projects such as <span class="error">&#91;kafka-storm-starter&#93;</span>(<a href="https://github.com/miguno/kafka-storm-starter" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/miguno/kafka-storm-starter</a>). </p>

<p>    Do you think we should add this to the Storm docs?  (I took me hours to track this issue down, for instance, and only after I found a fix myself did I read your "Heureka" comment in one of the Storm issues/pull requests.)</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/121#issuecomment-50685964" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/121#issuecomment-50685964</a></p>

<p>    Yes, lets put it in the documentation.  For most systems a 1 second delay every so often is not a big deal, but with storm being near real time it can be a big impact., not just for tests.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/121#issuecomment-62754593" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/121#issuecomment-62754593</a></p>

<p>    I think we are at the point where we should merge the security branch into master.  This is a little bit different from most patches/pull requests we have done, partly because of the size of it, and partly because of the large number of contributions from different individuals.  All of the code changes individually have been reviewed by storm committers before being pulled into this branch, but I would really like to see at least 2 committers besides myself give a +1 to pulling this in.  I think it would be great to have more people look at it, but unless they are blocking issues I would prefer that we file follow up JIRA for them, and get this in so I don't have to keep maintaining it.</p>

<p>    I am +1.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user harshach commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/121#issuecomment-62758258" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/121#issuecomment-62758258</a></p>

<p>    I am +1. Feedback on overall security branch we tested it over a period of time with these merges in with master and there aren't any issues for storm regular and storm security.<br/>
    @revans2  one question do we need to wait for 0.9.3 release to go out ?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/121#issuecomment-62782715" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/121#issuecomment-62782715</a></p>

<p>    Like I said in the mailing list, I have been out for a few weeks.  I didn't see 0.9.3 go out, but I thought that it must be very close by now.  I'll send a separate mail to the list to check on that status.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user ptgoetz commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/121#issuecomment-62839886" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/121#issuecomment-62839886</a></p>

<p>    I also replied to the list, I'm +1 for merging to master once we create a 0.9.3 branch.</p>

<p>    The modifications in the security branch have been QA'd fairly extensively at hortonworks and yahoo. I'm comfortable with moving forward with this work provided we have process in place to support 0.9.3.x releases.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/121#issuecomment-62944630" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/121#issuecomment-62944630</a></p>

<p>    I see a 0.9.3-branch already out in the git repo.  I assume that this is good enough.  I will merge in the security work but just to be sure I'll tag the repo right before the merge, in case we need to revert it for whatever reason.  Thanks everyone for all of your help.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user ptgoetz commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/121#issuecomment-62946641" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/121#issuecomment-62946641</a></p>

<p>    @revans2 Yeah, I created that branch earlier today then got distracted by meetings before I could send out a notification. That branch is currently identical to master.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2">revans2</a>:</i>
<p>This was merged into master.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user asfgit closed the pull request at:</p>

<p>    <a href="https://github.com/apache/storm/pull/121" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/121</a></p>
